### ISSUE TITLE
Access to instance attribute with callable type causes error

### ISSUE THREAD

**Issue by unknownlighter** [python/mypy#6910]
test_dc.py:
```
from dataclasses import dataclass, fields


@dataclass
class Foo:
    value: str


fields(Foo)[0].default_factory
```

```
$ mypy test_dc.py 
test_dc.py:9: error: Attribute function "default_factory" with type "Callable[[], Any]" does not accept self argument
```

python version: `3.7.1`
mypy version: `0.710+dev.95a04913df54d5b1ff78fcb9e7467b798491ec4b`

**Comment by ilevkivskyi** [comment]
It is not something specific to dataclasses and in fact may be tightly related to https://github.com/python/mypy/issues/708.

**Comment by ilevkivskyi** [comment]
Here is a minimal repro:
```python
from typing import Callable

class C:
    x: Callable[[], None]

x: C
x.x
```

**Comment by Strilanc** [comment]
Ran into this independently. Confirmed the repro still occurs in `mypy-0.740`.

**Comment by frou** [comment]
I notice that this works:

```python
class Thing(NamedTuple):
    f: Callable[[], int]

Thing(lambda: 123).f()
```

...and this doesn't:

```python
@dataclass
class Thing:
    f: Callable[[], int]

Thing(lambda: 123).f()

# error: Attribute function "f" with type "Callable[[], int]" does not accept self argument  [misc]
```

**Comment by sinback** [comment]
also stumbled across this one when checking whether a dataclasses.default_factory was an instance of dataclasses' _MISSING_TYPE value

**Comment by pbabics** [comment]
Hello I have run into this when using `enum.Enum`:

```python
import enum
class TestEnum2(enum.Enum):
	A = 1
	B = 2

for obj in TestEnum2:
	print(obj)
```

Result:
```
test.py:9: error: Attribute function "__iter__" with type "Callable[[], _EnumMetaType]" does not accept self argument
test.py:9: error: Attribute function "__next__" with type "Callable[[], TestEnum2]" does not accept self argument
```

Tested on:
`mypy 0.770+dev.c90b405934cbf2142b8a575ddd89dbedf762106c` (current master)
and also `mypy 0.770` release

**Comment by JelleZijlstra** [comment]
@pbabics I can't reproduce that error. Your line numbers don't match your code sample, so maybe you were typechecking a different file.

**Comment by pbabics** [comment]
Yes, you are right, I was testing this with our internal plugin that shadows standard Enum type checking, sorry for confusion

**Comment by staticdev** [comment]
@JelleZijlstra the error still happens in 0.780 and latest if you try on https://mypy-play.net/?mypy=latest&python=3.8
 
```from typing import Callable

class C:
    x: Callable[[], None]

x: C
x.x
```

**Comment by snejus** [comment]
Can confirm this is still happening with a `dataclass` on `0.782`.

**Comment by 9999years** [comment]
Well, if you don't want to wait for a fix in mypy, here's an extremely shitty workaround:

```python
from typing import Callable, cast

class C:
    _x: Callable[[], None]

    def x(self) -> None:
        return cast(Callable[[], None], getattr(self, "_x"))()
```

`getattr(self, "_x")` prevents a "redundant cast to `Callable[[], None]`" warning, and the cast is required to avoid passing the `self` argument to `self._x`.

This type-checks, but it's far from pretty.

**Comment by dargueta** [comment]
Is there a way to work around this for dataclasses? I have the following:


```py
import dataclasses


@dataclasses.dataclass
class A:
    x: float = dataclasses.field(default_factory=float)

fields = dataclasses.fields(A)
assert fields[0].default_factory is not dataclasses.MISSING
```

Running MyPy 0.812 with all default settings gives 

```
test.py:9: error: Attribute function "default_factory" with type "Callable[[], _T]" does not accept self argument  [misc]
Found 1 error in 1 file (checked 1 source file)
```

I would _really_ rather not add `# type: ignore[misc]` everywhere we do something like this, and @9999years' fix won't work for us because we're using [`desert`](https://desert.readthedocs.io/en/stable/) in conjunction with it and that would change the setting's name.

**Comment by hauntsaninja** [comment]
The issue with default_factory was fixed by a typeshed change (https://github.com/python/mypy/issues/10750 , https://github.com/python/typeshed/pull/5718). The underlying logic still needs fixing though, i.e., the min repro here still reproes https://github.com/python/mypy/issues/6910#issuecomment-497985882

If you're running into this, a callback protocol is a good workaround (see https://github.com/python/typeshed/pull/5718 for an example).

**Comment by ilevkivskyi** [comment]
As @hauntsaninja mentioned original example was already fixed, the other example is also working now. Probably was fixed by the same PR that fixed #708
